chain/chain/src/flat_storage_resharder.rs

```diff
@@ -0,0 +1,943 @@
+            FlatStorageReshardingStatus::CreatingChild => {
+                // Nothing to do here because the parent will take care of resuming work.
+            }
+            FlatStorageReshardingStatus::SplittingParent(status) => {
+                let parent_shard_uid = shard_uid;
+                info!(target: "resharding", ?parent_shard_uid, ?status, "resuming flat storage shard split");
+                self.check_no_resharding_in_progress()?;
+                // On resume flat storage status is already set.
+                // However, we don't know the current state of children shards,
+                // so it's better to clean them.
+                self.clean_children_shards(&status)?;
+                self.schedule_split_shard(parent_shard_uid, &status, scheduler, controller);
+            }
+            FlatStorageReshardingStatus::CatchingUp(_) => {
+                info!(target: "resharding", ?shard_uid, ?status, "resuming flat storage shard catchup");
+                // TODO(Trisfald): implement child catch up
+                todo!()
+            }
+        }
+        Ok(())
+    }
+
+    /// Starts a resharding event.
+    ///
+    /// For now, only splitting a shard is supported.
+    ///
+    /// # Args:
+    /// * `event_type`: the type of resharding event
+    /// * `shard_layout`: the new shard layout
+    /// * `scheduler`: component used to schedule the background tasks
+    /// * `controller`: manages the execution of the background tasks
+    pub fn start_resharding(
+        &self,
+        event_type: ReshardingEventType,
+        shard_layout: &ShardLayout,
+        scheduler: &dyn FlatStorageResharderScheduler,
+        controller: FlatStorageResharderController,
+    ) -> Result<(), Error> {
+        match event_type {
+            ReshardingEventType::SplitShard(params) => {
+                self.split_shard(params, shard_layout, scheduler, controller)
+            }
+        }
+    }
+
+    /// Starts the event of splitting a parent shard flat storage into two children.
+    fn split_shard(
+        &self,
+        split_params: ReshardingSplitShardParams,
+        shard_layout: &ShardLayout,
+        scheduler: &dyn FlatStorageResharderScheduler,
+        controller: FlatStorageResharderController,
+    ) -> Result<(), Error> {
+        let ReshardingSplitShardParams {
+            parent_shard,
+            left_child_shard,
+            right_child_shard,
+            block_hash,
+            prev_block_hash,
+            ..
+        } = split_params;
+        info!(target: "resharding", ?split_params, "initiating flat storage shard split");
+        self.check_no_resharding_in_progress()?;
+
+        // Change parent and children shards flat storage status.
+        let store = self.runtime.store().flat_store();
+        let mut store_update = store.store_update();
+        let flat_head = retrieve_shard_flat_head(parent_shard, &store)?;
+        let status = SplittingParentStatus {
```

Comment:
nit: I know I might have suggested the status name but it's strange to me that status is a static thing that doesn't actually reflect the progress of resharding. On top of that it seems very similar to params. Is there any chance that you can replace it with just params? Or maybe make the value of FlatStorageReshardingStatus a tuple (params, layout, flat_head)?

Reply:
Yes, there's no actual evolving status in this struct, although it might change in the future if we make flat storage resharding resume from the interruption point.

The difference between this and params is that this struct is specific to flat storage (it contains `flat_head`) whilst params is generic to resharding.

That said, I'll think about a better name